File: <100/065/065.cpp
#include<iostream>
#include<vector>
#include<numeric>
#include<boost/multiprecision/cpp_int.hpp>
#include<time.h>
using namespace std;
// namespace mp = boost::multiprecision;
using Bint   = boost::multiprecision::cpp_int;
using P      = pair<Bint, Bint>;


Bint digitSum(Bint n){
  if(n == 0) return (Bint)0;
  else       return n%(Bint)10 + digitSum(n/(Bint)10);
}


P reproduce_continued_fraction(vector<int> &V){
  Bint n = 0, d = 1;
  Bint N = V.size();

  while(V.size()){
    Bint tmp = d;
    Bint k = V.back();
    d = d*V.back() + n;
    V.pop_back();
    n = tmp;
    Bint hcf = gcd(n, d);
    n /= hcf, d /= hcf;
  }

  swap(n, d);
  return {n, d};
}


vector<int> generate_E(int sz){
  vector<int> V;
  if(sz == 0) return V;
  V.push_back((int)2);
  if(sz == 1) return V;
  V.push_back((int)1);
  if(sz == 2) return V;
  V.push_back((int)2);
  if(sz == 3) return V;

  int i = 1, j = 2;
  while((int)V.size() < sz){
    if(i > 0) V.push_back((int)1);
    else{
      V.push_back((int)2*j);
      ++j;
    }

    i = (i+1) % 3;
  }

  return V;
}


Bint solve(){
  vector<int> V = generate_E(100);

  P p = reproduce_continued_fraction(V);
  Bint answer = digitSum(p.first);
  return answer;
  return 1;
}

int main(){
  clock_t start = clock();
  Bint answer = solve();
  clock_t end = clock();
  const double time = static_cast<double>(end - start) / CLOCKS_PER_SEC;

  cout << answer << endl;
  cout << time << " sec" << endl;
}

File: <100/065/065.jl
function digitSum(n :: BigInt) :: BigInt
  result = 0
  while 0 < n
    result += n % 10
    n ÷= 10
  end

  return result
end


function reproduce_continued_fraction(v :: Vector{Int}) :: Tuple{BigInt, BigInt}
  n :: BigInt = 0
  d :: BigInt = 1
  N :: BigInt = length(v)

  while v |> isempty |> !
    tmp :: BigInt = d
    k :: BigInt = v[end]
    d = d * k + n
    n = tmp
    pop!(v)
    hcf = BigInt(gcd(n, d))
    n ÷= hcf
    d ÷= hcf
  end

  n, d = d, n
  return (n, d)
end


function generate_E(sz :: Int) :: Vector{Int}
  v :: Vector{Int} = []
  if sz == 0
    return v
  end
  push!(v, 2)
  if sz == 1
    return v
  end
  push!(v, 1)
  if sz == 2
    return v
  end
  push!(v, 2)
  if sz == 3
    return v
  end

  i :: Int = 1
  j :: Int = 2
  while length(v) < sz
    if i > 0
      push!(v, 1)
    else
      push!(v, 2*j)
      j += 1
    end

    i = (i+1) % 3
  end
  
  return v
end


function main() :: Nothing
  upper_limit :: Int = 100
  upper_limit |> generate_E |> reproduce_continued_fraction |> (t -> t[1]) |> digitSum |> println
end


@time main()
